nixos/modules/services/networking/networkmanager.nix

```diff
@@ -379,6 +380,52 @@ in {
           for more details.
         '';
       };
+      ensure-profiles = {
+        profiles = mkOption {
+          type = types.attrsOf ini.type;
```

Comment:
I wonder if there are some fields which are definitely mandatory for a profile to work and if we can add assertions that those are present.

Reply:
In my opinion the thing that would make the most sense to check is:
```nix
      connection = {
          id = "realraum";
          type = "wifi";
          interface-name = "wlp3s0";
        };
```
I think checking if connection exits and has the id attribute makes sense, everything else seems like a lot of work for something that's not really specified.

Adding checks for wifi, ipv4, ipv6, gsm, etc just seems overkill. There is also networkamanger plugins that add custom types and and you can just create as many interfaces adhoc as you want so checking interface-name and type also doesn't make too much sense in my opinion

Reply:
Do you have a suggestion on how I would implement this assertion?

Reply:
I'm thinking of something like this:
```
      {
        assertion = (forEach(mapAttrsToList(n: v: { inherit v; }) cfg.ensure-profiles.profiles) (profile: builtins.hasAttr "connection" profile));
        message = ''
          Every NetworkManager profile needs at least a connection setting with a id attribute
        '';
      }
```

Reply:
I agree that the assert will be useful.

Reply:
Do you have a implementation suggestion?

Reply:
Let's say something like this:
```
let
    isValidWiFiProfile =  ...;
    isValidConnection = connection: profile: hasAttr "id" connection && hasAttr "type" connection && (connection.type == "wifi" -> isValidWiFiProfile profile);
    isValidProfile = profile: (hasAttr "connection" profile && isValidConnection profile.connection profile)
```

And hop, we realize that we are reinventing submodules type checking. So let's go for a submodule here with freeform type options.

Then, the asserts should only be on "if this is a WiFi connection, this should have those fields minimally", etc.

Reply:
just massage it pre serialization

Reply:
look over all children and if you find that all subvalues are nulls, skip the section

Reply:
this should do the trick https://github.com/NixOS/nixpkgs/pull/254647/files#diff-0a708e7b053cf5df7620b5262936553af2242d2ce9dabde5bbeba221ece0a021R402 of filtering out empty attrsets now I have to figure out how to do the wifi submodule thing so it doesn't do the checks if it defaults to empty attrSet